x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java

```diff
@@ -0,0 +1,134 @@
+        Source source,
+        LogicalPlan child,
+        Expression tableName,
+        List<NamedExpression> matchFields,
+        @Nullable LocalRelation localRelation
+    ) {
+        super(source, child);
+        this.tableName = tableName;
+        this.matchFields = matchFields;
+        this.localRelation = localRelation;
+    }
+
+    public Lookup(PlanStreamInput in) throws IOException {
+        super(in.readSource(), in.readLogicalPlanNode());
+        this.tableName = in.readExpression();
+        this.matchFields = in.readCollectionAsList(i -> ((PlanStreamInput) i).readNamedExpression());
+        this.localRelation = in.readBoolean() ? new LocalRelation(in) : null;
+    }
+
+    public void writeTo(PlanStreamOutput out) throws IOException {
+        out.writeSource(source());
+        out.writeLogicalPlanNode(child());
+        out.writeExpression(tableName);
+        out.writeCollection(matchFields, (o, v) -> ((PlanStreamOutput) o).writeNamedExpression(v));
+        if (localRelation == null) {
+            out.writeBoolean(false);
+        } else {
+            out.writeBoolean(true);
+            localRelation.writeTo(out);
+        }
+    }
+
+    public Expression tableName() {
+        return tableName;
+    }
+
+    public List<NamedExpression> matchFields() {
+        return matchFields;
+    }
+
+    public LocalRelation localRelation() {
+        return localRelation;
+    }
+
+    @Override
+    public boolean expressionsResolved() {
+        return tableName.resolved() && Resolvables.resolved(matchFields);
+    }
+
+    @Override
+    public UnaryPlan replaceChild(LogicalPlan newChild) {
+        return new Lookup(source(), newChild, tableName, matchFields, localRelation);
+    }
+
+    @Override
+    protected NodeInfo<? extends LogicalPlan> info() {
+        return NodeInfo.create(this, Lookup::new, child(), tableName, matchFields, localRelation);
+    }
+
+    @Override
+    public List<Attribute> output() {
+        if (lazyOutput == null) {
+            List<? extends NamedExpression> rightSide = localRelation != null
+                ? Join.makeNullable(Join.makeReference(localRelation.output()))
+                : matchFields;
+            lazyOutput = mergeOutputAttributes(child().output(), rightSide);
+        }
+        return lazyOutput;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        if (super.equals(o) == false) {
+            return false;
+        }
+        Lookup lookup = (Lookup) o;
+        return Objects.equals(tableName, lookup.tableName)
+            && Objects.equals(matchFields, lookup.matchFields)
+            && Objects.equals(localRelation, lookup.localRelation);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), tableName, matchFields, localRelation);
```

Comment:
Thought: Hashing the local relation every time could be expensive, no?

Since the local relation is uniquely identified by the table name, we _could_ consider the `localRelation` to just be part of `Lookup`'s state (resolved vs. unresolved table) - and instead hash the boolean `localRelation == null`, no?
Similarly for `equals`.